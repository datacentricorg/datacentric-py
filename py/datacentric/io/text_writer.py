# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from abc import ABC, abstractmethod


class TextWriter(ABC):
    """
    Represents a writer that can write a sequential series of characters.

    This class is abstract.
    """

    @abstractmethod
    def write(self, value: object) -> None:
        """
        Write __str__ of the argument to the output stream.
        """
        pass

    @abstractmethod
    def write_line(self, value: object) -> None:
        """
        Write __str__ of the argument to the output stream, followed by EOL.
        """
        pass

    @abstractmethod
    def flush(self) -> None:
        """
        Flush buffer to the output stream.
        """
        pass
