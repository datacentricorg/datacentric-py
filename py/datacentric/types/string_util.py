# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import re
from typing import Pattern


class StringUtil():
    """
    Static helper class for str.
    """

    __first_cap_re: Pattern = re.compile('(.)([A-Z][a-z]+)')
    __all_cap_re: Pattern = re.compile('([a-z0-9])([A-Z])')

    @staticmethod
    def to_pascal_case(name: str) -> str:
        """
        Converts strings to PascalCase also removing underscores.
        """
        result: str = ''.join(x for x in name.title() if not x == '_')
        return result

    @staticmethod
    def to_snake_case(name: str) -> str:
        """
        Converts PascalCase strings to snake_case.
        """
        s1: str = StringUtil.__first_cap_re.sub(r'\1_\2', name)
        result: str = StringUtil.__all_cap_re.sub(r'\1_\2', s1).lower()
        return result
