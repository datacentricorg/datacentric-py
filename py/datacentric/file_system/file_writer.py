# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from datacentric.file_system.text_writer import TextWriter


class FileWriter(TextWriter):
    """
    Implements TextWriter for writing to a file.
    """

    __slots__ = ('__file',)

    # __file: - TODO - specify type

    def __init__(self, path: str):
        """
        Open the specified file for writing.

        The argument is full path to the file.
        """

        # Open in mode that overwrites the context
        self.__file = open(file=path, mode="w")

    def write(self, value: object) -> None:
        """
        Write __str__ of the argument to the output stream.
        """
        print(value, end = '', file=self.__file)

    def write_line(self, value: object) -> None:
        """
        Write __str__ of the argument to the output stream, followed by EOL.
        """
        self.write(value)
        self.write_eol()

    def write_eol(self) -> None:
        """
        Write EOL to the output stream.
        """
        print(file=self.__file)

    def flush(self) -> None:
        """
        Flush buffer to the output stream.
        """
        self.__file.flush()

    def __del__(self):
        """
        Close the file.
        """
        self.__file.close()
