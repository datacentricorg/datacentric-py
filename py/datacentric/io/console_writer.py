# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from datacentric.io.text_writer import TextWriter

class ConsoleWriter(TextWriter):
    """
    Implements TextWriter for writing to the console.
    """

    def write(self, value: object) -> None:
        """
        Write __str__ of the argument to the output stream.
        """
        print(value)

    def write_line(self, value: object) -> None:
        """
        Write __str__ of the argument to the output stream, followed by EOL.
        """
        self.write(value)
        print()

    def flush(self) -> None:
        """
        Flush buffer to the output stream.
        """

        # Do nothing
        pass