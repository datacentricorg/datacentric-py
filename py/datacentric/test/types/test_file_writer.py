# Copyright (C) 2013-present The DataCentric Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest
from datacentric.testing.unit_test import UnitTest
from datacentric.types.string_util import StringUtil
from datacentric.io.text_writer import TextWriter
from datacentric.io.file_writer import FileWriter

class TestStringUtil(unittest.TestCase, UnitTest):
    def test_smoke(self):
        """Smoke test"""
        file_path: str = __file__.replace(".py",".approved.txt")
        file_writer: TextWriter = FileWriter(file_path)

        # Does not remove space TODO - check that Humanizer does in C#
        file_writer.write_line(StringUtil.to_pascal_case('abc def'))
        file_writer.write_line(StringUtil.to_pascal_case('abc_def'))

        # Does not remove space TODO - check that Humanizer does in C#
        file_writer.write_line(StringUtil.to_snake_case('Abc Def'))
        file_writer.write_line(StringUtil.to_snake_case('AbcDef'))

if __name__ == "__main__":
    unittest.main()
