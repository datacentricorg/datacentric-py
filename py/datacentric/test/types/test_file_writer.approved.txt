Abc Def
AbcDef
abc _def
abc_def
